## Advent of Code

Advent of Code is an Advent calendar of small programming puzzles for a variety of skill sets and skill levels that can be solved in any programming language you like. People use them as a speed contest, interview prep, company training, university coursework, practice problems, or to challenge each other.

Visit: https://adventofcode.com/2021/about for more info

This workbook contains my work to solve the different problems from 2021

### Day 1 Pt 1

In [2]:
import pandas as pd
import numpy as np

In [3]:
df = pd.read_csv('AOC_list.csv')

In [11]:
df

Numbers  Inc/Dec
0         109      NaN
1         117      NaN
2         118      NaN
3          98      NaN
4         102      NaN
...       ...      ...
1995     6289      NaN
1996     6292      NaN
1997     6293      NaN
1998     6297      NaN
1999     6308      NaN

[2000 rows x 2 columns]

In [20]:
#goal: count the number of times a depth measurement increases
#from previous measurement

for i in df.index[1:2000]:
    if df['Numbers'][i-1] < df['Numbers'][i]:
        df['Inc/Dec'][i] = 'increasing'
    else:
        df['Inc/Dec'][i] = 'decreasing'

<ipython-input-20-530a32510736>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Inc/Dec'][i] = 'increasing'
C:\Users\jluczak\Anaconda3\lib\site-packages\pandas\core\indexing.py:670: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value)
<ipython-input-20-530a32510736>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Inc/Dec'][i] = 'decreasing'


In [24]:
#returns count of increasing
df[df['Inc/Dec'] == 'increasing'].count()

Numbers    1448
Inc/Dec    1448
dtype: int64

There are 1,448 instances of records increasing from the previous record.

### Day 1 Pt 2

In [28]:
#answer the same question, but based off a sliding window 
#of 3 measurements

df2 = pd.DataFrame(df['Numbers'])

In [32]:
df2['Sum'] = np.nan
df2['inc/dec'] = 'n/a'

In [33]:
df2

Numbers  Sum inc/dec
0         109  NaN     n/a
1         117  NaN     n/a
2         118  NaN     n/a
3          98  NaN     n/a
4         102  NaN     n/a
...       ...  ...     ...
1995     6289  NaN     n/a
1996     6292  NaN     n/a
1997     6293  NaN     n/a
1998     6297  NaN     n/a
1999     6308  NaN     n/a

[2000 rows x 3 columns]

In [40]:
#create window sum values and populated in df2

for i in df2.index[2:2000]:
    df2['Sum'][i] = (df2['Numbers'][i-2] + df2['Numbers'][i-1] +df2['Numbers'][i])

<ipython-input-40-af36e1b2687b>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2['Sum'][i] = (df2['Numbers'][i-2] + df2['Numbers'][i-1] +df2['Numbers'][i])


In [41]:
df2

Numbers      Sum inc/dec
0         109      NaN     n/a
1         117      NaN     n/a
2         118    344.0     n/a
3          98    333.0     n/a
4         102    318.0     n/a
...       ...      ...     ...
1995     6289  18864.0     n/a
1996     6292  18868.0     n/a
1997     6293  18874.0     n/a
1998     6297  18882.0     n/a
1999     6308  18898.0     n/a

[2000 rows x 3 columns]

In [42]:
#calc if window sum inc/dec
for i in df2.index[3:2000]:
    if df2['Sum'][i-1] < df2['Sum'][i]:
        df2['inc/dec'][i] = 'increasing'
    else:
        df2['inc/dec'][i] = 'decreasing'

<ipython-input-42-84723f99360d>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2['inc/dec'][i] = 'decreasing'
<ipython-input-42-84723f99360d>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2['inc/dec'][i] = 'increasing'


In [43]:
df2

Numbers      Sum     inc/dec
0         109      NaN         n/a
1         117      NaN         n/a
2         118    344.0         n/a
3          98    333.0  decreasing
4         102    318.0  decreasing
...       ...      ...         ...
1995     6289  18864.0  increasing
1996     6292  18868.0  increasing
1997     6293  18874.0  increasing
1998     6297  18882.0  increasing
1999     6308  18898.0  increasing

[2000 rows x 3 columns]

In [45]:
df2[df2['inc/dec'] == 'increasing'].count()

Numbers    1471
Sum        1471
inc/dec    1471
dtype: int64

The resulting answer is 1,471 instances where the window sum of 3 numbers is increasing

### Day 2 Pt 1

In [3]:
df_2 = pd.read_csv('aoc_2.csv')

In [4]:
df_2

Directions  Distance
0      forward         2
1         down         4
2         down         3
3           up         4
4         down         1
..         ...       ...
995    forward         4
996         up         9
997       down         1
998       down         7
999    forward         1

[1000 rows x 2 columns]

In [8]:
horizontal = 0
vertical = 0

for i in df_2.index:
    if df_2['Directions'][i] == 'forward':
        horizontal += df_2['Distance'][i]
    elif df_2['Directions'][i] == 'down':
        vertical += df_2['Distance'][i]
    elif df_2['Directions'][i] == 'up':
        vertical -= df_2['Distance'][i]
        

In [9]:
horizontal

1906

In [10]:
vertical

1017

In [11]:
horizontal*vertical

1938402

The final answer is 1,938,402

### Day 2 Pt 2

In [12]:
df_2

Directions  Distance
0      forward         2
1         down         4
2         down         3
3           up         4
4         down         1
..         ...       ...
995    forward         4
996         up         9
997       down         1
998       down         7
999    forward         1

[1000 rows x 2 columns]

In [53]:
horizontal = 0
depth = 0
aim = 0

print("Starting horizontal {}".format(horizontal))
print("Starting depth {}".format(depth))
print("Starting aim {}".format(aim))
print()

for i in df_2.index:
    print("Index {}".format(i))
    print("Direction {}".format(df_2['Directions'][i]))
    print("Distance {}".format(df_2['Distance'][i]))
    print()
    if df_2['Directions'][i] == 'forward':
        horizontal += df_2['Distance'][i]
        print('horizontal position now {}'.format(horizontal))
        if aim > 0:
            depth = depth + (aim * df_2['Distance'][i])
            print("depth is now {}".format(depth))
        print('aim now {}'.format(aim))
        print("depth is now {}".format(depth))
    elif df_2['Directions'][i] == 'down':
        aim += df_2['Distance'][i]
        print("aim is now {}".format(aim))
    elif df_2['Directions'][i] == 'up':
        aim -= df_2['Distance'][i]
        print("aim is now {}".format(aim))

Starting horizontal 0
Starting depth 0
Starting aim 0

Index 0
Direction forward
Distance 2

horizontal position now 2
aim now 0
depth is now 0
Index 1
Direction down
Distance 4

aim is now 4
Index 2
Direction down
Distance 3

aim is now 7
Index 3
Direction up
Distance 4

aim is now 3
Index 4
Direction down
Distance 1

aim is now 4
Index 5
Direction down
Distance 8

aim is now 12
Index 6
Direction up
Distance 9

aim is now 3
Index 7
Direction forward
Distance 1

horizontal position now 3
depth is now 3
aim now 3
depth is now 3
Index 8
Direction down
Distance 9

aim is now 12
Index 9
Direction forward
Distance 6

horizontal position now 9
depth is now 75
aim now 12
depth is now 75
Index 10
Direction down
Distance 7

aim is now 19
Index 11
Direction forward
Distance 1

horizontal position now 10
depth is now 94
aim now 19
depth is now 94
Index 12
Direction down
Distance 1

aim is now 20
Index 13
Direction up
Distance 2

aim is now 18
Index 14
Direction forward
Distance 8

horizontal posi

aim now 551
depth is now 216494
Index 448
Direction up
Distance 6

aim is now 545
Index 449
Direction down
Distance 4

aim is now 549
Index 450
Direction forward
Distance 2

horizontal position now 926
depth is now 217592
aim now 549
depth is now 217592
Index 451
Direction up
Distance 3

aim is now 546
Index 452
Direction down
Distance 2

aim is now 548
Index 453
Direction up
Distance 1

aim is now 547
Index 454
Direction up
Distance 7

aim is now 540
Index 455
Direction down
Distance 5

aim is now 545
Index 456
Direction up
Distance 3

aim is now 542
Index 457
Direction forward
Distance 5

horizontal position now 931
depth is now 220302
aim now 542
depth is now 220302
Index 458
Direction up
Distance 2

aim is now 540
Index 459
Direction down
Distance 9

aim is now 549
Index 460
Direction down
Distance 6

aim is now 555
Index 461
Direction forward
Distance 5

horizontal position now 936
depth is now 223077
aim now 555
depth is now 223077
Index 462
Direction up
Distance 3

aim is now 55

Index 864
Direction up
Distance 4

aim is now 925
Index 865
Direction down
Distance 5

aim is now 930
Index 866
Direction up
Distance 1

aim is now 929
Index 867
Direction forward
Distance 2

horizontal position now 1615
depth is now 746736
aim now 929
depth is now 746736
Index 868
Direction forward
Distance 2

horizontal position now 1617
depth is now 748594
aim now 929
depth is now 748594
Index 869
Direction up
Distance 9

aim is now 920
Index 870
Direction forward
Distance 8

horizontal position now 1625
depth is now 755954
aim now 920
depth is now 755954
Index 871
Direction up
Distance 4

aim is now 916
Index 872
Direction up
Distance 1

aim is now 915
Index 873
Direction up
Distance 6

aim is now 909
Index 874
Direction up
Distance 5

aim is now 904
Index 875
Direction down
Distance 5

aim is now 909
Index 876
Direction up
Distance 4

aim is now 905
Index 877
Direction up
Distance 1

aim is now 904
Index 878
Direction forward
Distance 9

horizontal position now 1634
depth is now 7

In [54]:
horizontal*depth

1947878632

### Day 3 Part 1

In [72]:
df_3 = pd.read_excel('aoc_3.xlsx', dtype=str)

In [73]:
df_3

Value 1 Value 2 Value 3 Value 4 Value 5 Value 52 Value 7 Value 8 Value 9  \
0         0       1       1       1       1        0       0       1       1   
1         1       0       1       1       0        1       0       0       1   
2         0       0       0       0       0        0       0       1       0   
3         1       0       0       1       1        1       0       0       1   
4         1       1       0       0       0        0       0       1       1   
..      ...     ...     ...     ...     ...      ...     ...     ...     ...   
995       0       0       0       1       1        1       1       0       0   
996       0       0       0       0       1        1       1       0       1   
997       0       1       1       0       1        0       0       0       1   
998       0       0       1       0       1        1       0       1       0   
999       1       1       0       0       1        1       1       1       1   

    Value 10 Value 11 Value 12  
0          1        1        0  
1          1        1        1  
2          1        0        1  
3          0        1        0  
4          0        1        0  
..       ...      ...      ...  
995        1        1        0  
996        1        0        0  
997        0        0        0  
998        0        1        1  
999        0        1        1  

[1000 rows x 12 columns]

In [129]:
gamma = ''

for i in df_3.columns:
    gamma += str(df_3[i].mode()[0])
    
epsilon = ''
for i in range(len(gamma)):
    if gamma[i] == '0':
        epsilon += '1'
    elif gamma[i] == '1':
        epsilon += '0'   
    
gamma_dec = int(gamma,2)
epsilon_dec = int(epsilon,2)

answer = gamma_dec*epsilon_dec
print(answer)

775304


The answer is 775,304

### Day 3 Pt 2

In [406]:
#read data

df_3 = pd.read_excel('aoc_3.xlsx')
df_3

Value 1  Value 2  Value 3  Value 4  Value 5  Value 52  Value 7  Value 8  \
0          0        1        1        1        1         0        0        1   
1          1        0        1        1        0         1        0        0   
2          0        0        0        0        0         0        0        1   
3          1        0        0        1        1         1        0        0   
4          1        1        0        0        0         0        0        1   
..       ...      ...      ...      ...      ...       ...      ...      ...   
995        0        0        0        1        1         1        1        0   
996        0        0        0        0        1         1        1        0   
997        0        1        1        0        1         0        0        0   
998        0        0        1        0        1         1        0        1   
999        1        1        0        0        1         1        1        1   

     Value 9  Value 10  Value 11  Value 12  
0          1         1         1         0  
1          1         1         1         1  
2          0         1         0         1  
3          1         0         1         0  
4          1         0         1         0  
..       ...       ...       ...       ...  
995        0         1         1         0  
996        1         1         0         0  
997        1         0         0         0  
998        0         0         1         1  
999        1         0         1         1  

[1000 rows x 12 columns]

In [404]:
#for oxygen: if mode of values is 0 then return only records where the first bit is 0; same goes if 1
# if the value counts of both 0 and 1 in that bit are the same, grab records for 1

oxy  = df_3

for i in oxy.columns:
    if oxy[i].value_counts()[0] == oxy[i].value_counts()[1]:
            oxy = oxy[oxy[i] == '1']
    else:
        oxy = oxy[oxy[i] == oxy[i].mode()[0]]
    
oxy.reset_index(drop=True, inplace=True)

#concat the bits into one value that we can convert into a decimal
oxy_out = ''

for i in range(len(oxy.iloc[0])):
    oxy_out += oxy.iloc[0][i]
    
oxy_out = int(oxy_out, 2)

# for co2: grab the least common value in that bit and return records that equal that value
# if value counts of both 0 and 1 are equal, grab bits with entry of 0

co2 = df_3

for i in co2.columns:
    if len(co2) > 1:
        if co2[i].value_counts()[0] == co2[i].value_counts()[1]:
            co2 = co2[co2[i] == '0']
        else:
            co2 = co2[co2[i] == co2[i].value_counts().index.tolist()[1]]
    
co2.reset_index(drop=True, inplace=True)

co2_out = ''

for i in range(len(co2.iloc[0])):
    co2_out += co2.iloc[0][i]
    
co2_out = int(co2_out, 2)

#multiply to get final result

final_ans = oxy_out*co2_out
final_ans

1370737

### Day 4 Pt 1

In [3]:
numbers = pd.read_csv('aoc4.csv')

In [4]:
tables = pd.read_excel('aoc4_bingotables.xlsx')
tables

Col 1  Col 2  Col 3  Col 4  Col 5
0       88     67     20     19     15
1       22     76     86     44     73
2        7     42      6     69     25
3       12     68     92     21     75
4       97     45     13     52     70
..     ...    ...    ...    ...    ...
495     26     53     58     52      1
496     82     57     32     40     20
497     72     21     74     46     43
498     41     15     98      2     11
499      5     96     22     18     55

[500 rows x 5 columns]

In [5]:
test = tables
empty = test.copy()

for col in empty.columns:
    empty[col].values[:] = 0
empty

Col 1  Col 2  Col 3  Col 4  Col 5
0        0      0      0      0      0
1        0      0      0      0      0
2        0      0      0      0      0
3        0      0      0      0      0
4        0      0      0      0      0
..     ...    ...    ...    ...    ...
495      0      0      0      0      0
496      0      0      0      0      0
497      0      0      0      0      0
498      0      0      0      0      0
499      0      0      0      0      0

[500 rows x 5 columns]

Theory of how to complete: convert the table into binary elements. if record exists then becomes a 1 and stays that way as we iterate through. if some of row or sum of column = 5 then it is a winner

Iterate through numbers list and then update the table each time, grab sums, if no winner then continue

In [11]:
#store the output of number of row sum and column sum to a dataframe
# so we can check at the end whether there is a winner
col_match = pd.DataFrame(index=range(0,500,5), 
                         columns=['Col 1', 'Col 2', 'Col 3',
                                 'Col 4', 'Col 5'])
bingo = 0

for x in numbers.index:
#         print('index val', x)
        r = 0
#         print('{} was marked off cards'.format(numbers.iloc[x][0]))
        for i in test.columns:
            q = 0
            last_num = 1000
            for z in test[i].index:
            #if the value equals number called set binary value to 1
                if test[i][z] == numbers.iloc[x][0]:
#                     print(i, 'loop 1 - mark numbers')
                    empty[i][z] = 1
                else:
                    continue
                    
            while q <= len(test):
                #sum up groups of 5 rows
                col_sum = sum(empty[i].iloc[q:q+5])
#             if col_sum == 5:
# #             #if we get bingo we stop the game
#                 print(i, 'loop 2 - sum up columns')
#                 break
#                 print(i)
                col_match[i].loc[q] = col_sum
#                 print('values added to column DF')
                q +=5
        
        for i in test.columns:
    #testing columns or any bingos
            if col_match[i][col_match[i] == 5].count() > 0:
                print('Bingo!')
                print('The last number called for bingo is {}'.format(numbers.iloc[x][0]))
                last_num = numbers.iloc[x][0]
                bingo = 1
                curr_ind = col_match[i][col_match[i] == 5].index[0]
                break

        while r in test.index:
#             print(r)
#             print('loop 3 - sum up rows')
#             print('checking row sums')
            row_sum = np.sum(empty.iloc[r])
            if row_sum == 5:
                print('Bingo!')
                print('The last number called for bingo is {}'.format(numbers.iloc[x][0]))
                last_num = numbers.iloc[x][0]
                curr_ind = test.iloc[r]
                bingo = 1
                break

#             #if we get bingo we stop the game
            r +=1
    
        if bingo == 1:
            break
            


Bingo!
The last number called for bingo is 78


In [7]:
total_table = empty[curr_ind:curr_ind+5]
tot_list = list()

for i in total_table.columns:
    for z in total_table.index:
        if total_table[i].loc[z] == 0:
            tot_list.append(test[i].loc[z])
            
tot = sum(tot_list)
output = tot*last_num
output

49686

The answer is 49686

### Day 4 Pt 2

In [183]:
test = tables
empty = test.copy()

for col in empty.columns:
    empty[col].values[:] = 0
empty

Col 1  Col 2  Col 3  Col 4  Col 5
0        0      0      0      0      0
1        0      0      0      0      0
2        0      0      0      0      0
3        0      0      0      0      0
4        0      0      0      0      0
..     ...    ...    ...    ...    ...
495      0      0      0      0      0
496      0      0      0      0      0
497      0      0      0      0      0
498      0      0      0      0      0
499      0      0      0      0      0

[500 rows x 5 columns]

In [185]:
#need to return the last winning bingo card
#to do that need to store winning rows and columns by card into new df and then update dictionary anytime a new card
#gets bingo

#instantiate variables
col_match = pd.DataFrame(index=range(0,500,5), 
                         columns=['Col 1', 'Col 2', 'Col 3',
                                 'Col 4', 'Col 5'])
row_match = pd.DataFrame(index=range(0,500,5), 
                         columns=['Row'])

row_list = list()
winner_list = list()
winner_dict = dict()


for x in numbers.index:
        r = 0
        for i in test.columns:
            q = 0
            last_num = 1000
            for z in test[i].index:
            #loops through all bingo cards and sets value of 1 if the number called matches a value in that bingo card
                if test[i][z] == numbers.iloc[x][0]:
                    empty[i][z] = 1
                else:
                    continue
                    
            while q <= len(test):
                #sum up the rows from each column and save to a dataframe so we can check did this card have 5 in a row
                col_sum = sum(empty[i].iloc[q:q+5])
                col_match[i].loc[q] = col_sum
                q +=5

        while r in test.index:
            #same thing for rows. Did this card match 5 across the rows
            row_sum = np.sum(empty.iloc[r])
            row_list.append(row_sum)
            if len(row_list) == 5:
                row_match['Row'].loc[r-4] = max(row_list)
                row_list = list()

            r +=1
    
    #create a new dataframe that joins column and row matches so we can check each index whether it won
        winner_test = pd.concat([col_match, row_match], axis=1)
        for z in winner_test.index:
            for a in winner_test.columns:
                winner_list.append(winner_test[a].loc[z])
                if len(winner_list) == 6:
                    #allows us to re-count after each card
                    if 5 in winner_list:
                        #anytime there is a new winner it overwrites and maintains order of winners
                        winner_dict.update({z:'win'})
                    winner_list = list()
                
        if len(winner_dict) == 100:
            #once all 100 cards win, we stop the game and store the last number called
            last_num = numbers.iloc[x][0]
            break
            

In [198]:
dict_list = list(winner_dict.keys())

#grab the data from the last winning card
total_table2 = empty[dict_list[99]:dict_list[99]+5]
tot_list2 = list()

for i in total_table2.columns:
    for z in total_table2.index:
        #sum up al records where we didn't match a called number
        if total_table2[i].loc[z] == 0:
            tot_list2.append(test[i].loc[z])
            
tot = sum(tot_list2)
output = tot*last_num
output

26878

### Day 5 Pt 1

In [200]:
df = pd.read_excel('aoc_5.xlsx')

In [201]:
df

x1   y1   x2   y2
0    223  805  223  548
1    609  164  609  503
2    461  552  796  552
3    207  361  207   34
4    503  879  503  946
..   ...  ...  ...  ...
495  387  601  387  597
496  355  186  782  613
497  866  435  816  435
498   96  161  764  161
499  971   29   21  979

[500 rows x 4 columns]

### Day 6 Pt 1

In [202]:
lf = pd.read_csv('aoc_6.csv')

In [203]:
lf

Fish
0       5
1       1
2       2
3       1
4       5
..    ...
295     1
296     1
297     4
298     1
299     5

[300 rows x 1 columns]

Theory of how to solve:

For loop that simulates the days 0:79 (80 days)

Each day we subtract a value from the vaue in the dataframe
once it reaches 0 we do not allow it to go to -1, but reset to 6

When value is reset a new fish is created with value of 8
may need a list to do this

Keep going until we reach end